### PR TITLE
Do not generate nil_exprt in bv_get

### PR DESCRIPTION
**Pull request by tautschnig**
We have all the information available, and should thus always generate concrete
values.

This is an attempt to address what has been discussed in #4153, but completely missing automated tests (I've done some local hacks on my L1-renaming branch). @thk123 would you be able to take a look and possibly assist with tests? 

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/solvers/flattening/boolbv_get.cpp [hunk 1]
```diff
@@ -143,15 +144,14 @@ exprt boolbvt::bv_get_rec(
       for(const auto &c : components)
       {
         const typet &subtype = c.type();
-        op.push_back(nil_exprt());
 
-        std::size_t sub_width=boolbv_width(subtype);
+        const member_exprt member(expr, c.get_name(), subtype);
+        op.push_back(
```
Comment (kroening) [inline src/solvers/flattening/boolbv_get.cpp:157]: Doesn't this then produce a member of nil?

File: src/solvers/flattening/boolbv_get.cpp [hunk 2]
```diff
@@ -168,9 +168,11 @@ exprt boolbvt::bv_get_rec(
 
       const typet &subtype = components[component_nr].type();
 
+      const member_exprt member(
+        expr, components[component_nr].get_name(), subtype);
       return union_exprt(
```
Comment (kroening) [inline src/solvers/flattening/boolbv_get.cpp:181]: same here?

File: src/solvers/flattening/boolbv_get.cpp [hunk 3]
```diff
@@ -262,14 +268,14 @@ exprt boolbvt::bv_get_rec(
   }
   }
 
-  return nil_exprt();
+  UNREACHABLE;
```
Comment (kroening) [inline src/solvers/flattening/boolbv_get.cpp:285]: There should be a PRECONDITION in the if(...) above on the type, or a failure here will be hard to debug.

Comment (tautschnig) [reply src/solvers/flattening/boolbv_get.cpp:285]: ~I have borrowed a bit from #2548.~ (Edit: sorry, that doesn't actually help here, I've added the appropriate `PRECONDITION` instead.)

File: src/solvers/flattening/boolbv_get.cpp (hunk 1 above)
Comment (tautschnig) [reply src/solvers/flattening/boolbv_get.cpp:157]: That was me being lazy, which I shouldn't have been. Now passing a proper `index_exprt`.

File: src/solvers/flattening/boolbv_get.cpp (hunk 2 above)
Comment (tautschnig) [reply src/solvers/flattening/boolbv_get.cpp:181]: As above, the `nil_exprt` is now an `index_exprt` and there can't be a `nil_exprt` reaching this point.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e9ffb73f7fdb48d7da855fc7dcb285261b32093e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106406254

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 14253f3a4bf6aa4fa2680d5a8ee889d0b05a6bf3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106420870

**Review by martin-cs — APPROVED** [review]
It seems like only the last commit is actually connected to the title of the PR.

**Comment by tautschnig** [comment]
> It seems like only the last commit is actually connected to the title of the PR.

Ah, sorry, I should have said so: the first two are from #4372, which I've marked this one as dependent-on.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7da0d6fdcbbbd8e49ff138c953149a841b1d1d48).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106870338

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 69c8b0d3950b25ed76827ffe2d178267718b1d65).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106929748

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 57874c448ab7c85e103ef6daedf7e68277f22b0d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106944413

**Pull request merged**